### PR TITLE
implement EnqueueExtensions in interPodAffinity

### PR DESCRIPTION
**Pull request by pravarag**
#### What type of PR is this?
/kind feature
/sig scheduling

#### What this PR does / why we need it:
Implement EnqueueExtensions interface in interPodAffinity plugins
#### Which issue(s) this PR fixes:

Part of #94009

### CODE REVIEW

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 1]
```diff
@@ -52,6 +53,15 @@ func (pl *InterPodAffinity) Name() string {
 	return Name
 }
 
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable
+func (pl *InterPodAffinity) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		{Resource: framework.Pod, ActionType: framework.Delete},
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:60]: PodAffinity is a bit different than other plugins, we should consider whether the event change can make a failed pod schedulable. So:

- PodDelete is one, b/c if a pod failed due to violating existing Pod's anti-affinity constraints, deleting an existing pods can make it schedulable
- PodUpdate can also be helpful, b/c an existing Pod's label can be changed (removed), so an unschedulable Pod may become schedulable
- PodAdd can also be helpful, b/c a unschedulable Pod may failed by pod-affinity constraints, and an added Pod may make it schedulable.

So we should put `framework.All` here.

**Comment by Huang-Wei** [comment]
@ahg-g I added some comments, but PodAffinity is a special plugin, so it'd be good if you can take a second look.

**Comment by ahg-g** [comment]
> @ahg-g I added some comments, but PodAffinity is a special plugin, so it'd be good if you can take a second look.

Will do, but lets not merge anymore this release?

/hold

**Comment by ahg-g** [comment]
> > Will do, but lets not merge anymore this release?
> 
> I'd like to let (most of) them in, but we can evaluate the risk case by case.

We need time to evaluate each filter, I would rather not rush everything today.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 1 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: I've replaced `framework.Pod` with `framework.All` as suggested.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 2]
```diff
@@ -52,6 +53,15 @@ func (pl *InterPodAffinity) Name() string {
...
+		{Resource: framework.Node, ActionType: framework.Add | framework.UpdateNodeAllocatable},
```
[the remark this answers is not in the document]
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: I've updated the framework to `UpdateNodeLabel` as suggested.

**Comment by Huang-Wei** [comment]
> Fixes #94009

@pravarag This would incidentally close #94009. Let's reword to "Part of #94009".

LGTM otherwise. Please squash the commits and update its "Draft" state.

**Comment by pravarag** [comment]
@Huang-Wei I've squashed the commits and edited the PR description as suggested above. Also, since I've not carried out the performance testing for the changes, will we be keeping the PR open till then or we can track those in the main issue itself?

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 3]
```diff
@@ -54,6 +55,15 @@ func (pl *InterPodAffinity) Name() string {
 	return Name
 }
 
+// EventsToRegister returns the possible events that may make a Pod
+// failed by this plugin schedulable
+func (pl *InterPodAffinity) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		{Resource: framework.Pod, ActionType: framework.All},
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:62]: Can you reword https://github.com/kubernetes/kubernetes/pull/100020#discussion_r590612561 here? As the reasons are not trivial, we'd better leave in-line comments here.

**Comment by Huang-Wei** [comment]
> Also, since I've not carried out the performance testing for the changes, will we be keeping the PR open till then or we can track those in the main issue itself?

Performance test is good-to-have, as we have enough time to let it soak in 1.22. You can remove the "Draft" state of this PR and it will be merged once the 1.22 cycle starts.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 3 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:64]: I've added inline comments as suggested with the latest push. Please let me know if anything else needs to be updated @Huang-Wei

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 4]
```diff
@@ -54,6 +55,17 @@ func (pl *InterPodAffinity) Name() string {
 	return Name
 }
 
+// EventsToRegister returns the possible events that may make a failed Pod
+// schedulable
+func (pl *InterPodAffinity) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		// All the events like PodDelete, PodUpdate, PodAdd can make a Pod schedulable
+		// so framework.All is being used here
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:63]:
```
// PodAll includes the following events:
// - PodDelete. An unschedulable Pod may fail due to violating an existing Pod's anti-affinity constraints, deleting an existing Pod may make it schedulable.
// - PodUpate. Updating on an existing Pod's labels (e.g., removal) may make an unschedulable Pod schedulable.
// - PodAdd. An unschedulable Pod may fail due to violating pod-affinity constraints, and an added Pod may make it schedulable.
```

Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: Updated the inline comments as mentioned above.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 5]
```diff
@@ -54,6 +55,19 @@ func (pl *InterPodAffinity) Name() string {
 	return Name
 }
 
+// EventsToRegister returns the possible events that may make a failed Pod
+// schedulable
+func (pl *InterPodAffinity) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		// PodAll includes the following events:
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:62]:
```suggestion
		// All ActionType includes the following:
```

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 6]
```diff
@@ -54,6 +55,19 @@ func (pl *InterPodAffinity) Name() string {
...
+		// - PodDelete. An unschedulable Pod may fail due to violating an existing Pod's anti-affinity constraints, deleting an existing Pod may make it schedulable.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:63]:
```suggestion
		// - Delete. An unschedulable Pod may fail due to violating an existing Pod's anti-affinity constraints,
		// deleting an existing Pod may make it schedulable.
```

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 7]
```diff
@@ -54,6 +55,19 @@ func (pl *InterPodAffinity) Name() string {
...
+		// - PodUpate. Updating on an existing Pod's labels (e.g., removal) may make an unschedulable Pod schedulable.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:64]:
```suggestion
		// - Update. Updating on an existing Pod's labels (e.g., removal) may make an unschedulable Pod 
		// schedulable.
```

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 8]
```diff
@@ -54,6 +55,19 @@ func (pl *InterPodAffinity) Name() string {
...
+		// - PodAdd. An unschedulable Pod may fail due to violating pod-affinity constraints, and an added Pod may make it schedulable.
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:65]:
```suggestion
		// - Add. An unschedulable Pod may fail due to violating pod-affinity constraints, adding an assigned
		// Pod may make it schedulable.
```

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go [hunk 9]
```diff
@@ -22,7 +22,7 @@ import (
 	"strings"
 	"testing"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
Comment (ahg-g) [inline pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:25]: revert the file.

**Comment by Huang-Wei** [comment]
/hold
@pravarag could you address some final comments from @ahg-g ?

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 5 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: Updated with latest push.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 6 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: Updated with latest push.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 7 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: Updated with latest push.

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go (hunk 8 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: Updated the review comment.

File: pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go (hunk 9 above)
Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:25]: @Huang-Wei @ahg-g this change, it's taking on it's own when I try to save the file. Did you mean revert the file to a previous commit?

Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go:25]: @ahg-g just meant to discard this change.

Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/scoring_test.go]: Done with the latest push

File: pkg/scheduler/framework/plugins/interpodaffinity/plugin.go [hunk 10]
```diff
@@ -54,6 +55,22 @@ func (pl *InterPodAffinity) Name() string {
...
 
+// EventsToRegister returns the possible events that may make a failed Pod
+// schedulable
+func (pl *InterPodAffinity) EventsToRegister() []framework.ClusterEvent {
+	return []framework.ClusterEvent{
+		// All ActionType includes the following events:
+		// - Delete. An unschedulable Pod may fail due to violating an existing Pod's anti-affinity constraints,
+		// deleting an existing Pod may make it schedulable.
+		// - Update. Updating on an existing Pod's labels (e.g., removal) may make
+		// an unschedulable Pod.
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/interpodaffinity/plugin.go:66]:
```suggestion
		// an unschedulable Pod schedulable.
```

Comment (pravarag) [reply pkg/scheduler/framework/plugins/interpodaffinity/plugin.go]: Updated with latest push.

**Pull request merged**